### PR TITLE
Add section for compiling non-BEAM code

### PR DESCRIPTION
**Pull request by fhunleth**
I couldn't find that we ever documented all of the environment variables that we set for building C/C++ code, so I started writing one.

### CODE REVIEW

File: docs/Compiling Non-BEAM Code.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+   `BusyBox init`. Initialization is done in either an
+   [Application.start callback](https://hexdocs.pm/elixir/Application.html#module-the-application-callback-module)
+   or in a `GenServer` so that it can be supervised.
+2. D-Bus is not normally enabled on Nerves. It may be enabled in a custom
+   system.
+3. X Windows is not used. Again, it may be enabled, but it is far more common to
+   have UI applications be fullscreen and not use a window manager.
+4. Minimal commands are available for shell scripts. Nerves ships with the
+   [`ash`](https://en.wikipedia.org/wiki/Almquist_shell) and bash scripts
+   nearly always require adjustments.
```
Comment (jjcarstens) [inline docs/Compiling Non-BEAM Code.md:36]: Im not quite sure what is trying to be said here? Nerves has `ash`, but bash requires adjustments to be used?

File: docs/Compiling Non-BEAM Code.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+2. Do not have a `priv` directory in your source tree. Instead copy static
+   assets into `$(MIX_APP_PATH)/priv` as part of the boot process
+3. If your library expects to write data to `priv`, provide an option to change
+   the location. On Nerves, the `priv` directory is read-only by design.
+4. Use `$(CC)` in your `Makefile` to compile for the target
+5. Use `$(CC_FOR_BUILD)` in your `Makefile` to compile tools for the host
+6. If using `elixir_make`, consider overriding the default error message since
+   it probably isn't right for cross-compilation
+
+Two recommendations cannot be stressed enough:
```
Comment (jjcarstens) [inline docs/Compiling Non-BEAM Code.md:146]:
```suggestion
Three recommendations cannot be stressed enough:
```

File: docs/Compiling Non-BEAM Code.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+First, always compile under `_build`. While it's much easier to compile in the
```
Comment (jjcarstens) [inline docs/Compiling Non-BEAM Code.md:148]:
```suggestion
1. Always compile under `_build`. While it's much easier to compile in the
```

File: docs/Compiling Non-BEAM Code.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+source directory, this always leads to errors where an executable compiled for
+one architecture (the host) ends up being put on the target. Nerves will fail
+with an error when this happens, but it causes a lot of confusion.
+
+Second, do not have a `priv` directory in your source tree. While Elixir
```
Comment (jjcarstens) [inline docs/Compiling Non-BEAM Code.md:153]:
```suggestion
2. Do not have a `priv` directory in your source tree. While Elixir
```

File: docs/Compiling Non-BEAM Code.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+provides a shortcut for copying files from a source `priv` directory to the
+build output `priv` directory, experience has been that this feature causes
+confusion when building native code. If you do have static assets that you want
+in the output `priv` directory, add a line to your `Makefile` or `mix.exs` to
+copy them.
+
+If you have the choice between using a NIF or a port to interface external code
```
Comment (jjcarstens) [inline docs/Compiling Non-BEAM Code.md:160]:
```suggestion
3. If you have the choice between using a NIF or a port to interface external code
```

**Comment by fhunleth** [comment]
@jjcarstens I ended up fixing or completely rewriting the areas that you commented on. They're a little different now. I think it's better. Thanks for the comments.

**Review by jjcarstens — APPROVED** [review]

**Pull request merged**